File: Swift/queue.swift
internal final class Queue<T> {

  internal var count: Int

  internal var isEmpty: Bool {
    get { return self.count == 0 }
  }

  internal var isFull: Bool {
    get { return self.count == self.items.count }
  }

  private var items: Array<T?>
  private var first: Int
  private var last: Int

  internal init(count: Int) {
    self.items = [T?](repeating: nil, count: count)
    self.last  = count - 1
    self.first = 0
    self.count = 0
  }

  internal func peek() throws -> T? {

    guard !self.isEmpty else {
      throw QueueError.emptyQueue(reason: "Cannot peek at an empty queue!")
    }

    return self.items[self.first]
  }

  internal func dequeue() throws -> T? {

    guard !self.isEmpty else {
      throw QueueError.emptyQueue(reason: "Cannot dequeue from an empty queue!")
    }

    let item: T? = self.items[self.first]
    self.items[self.first] = nil
    self.first = (self.first + 1) % self.items.count
    self.count -= 1
    return item
  }

  internal func enqueue(item: T?) throws -> Void {

    guard !self.isFull else {
      throw QueueError.fullQueue(reason: "Cannot enqueue into a full queue!")
    }

    self.last = (self.last + 1) % self.items.count
    self.items[self.last] = item
    self.count += 1
  }

}

internal enum QueueError: Error {
  case emptyQueue(reason: String)
  case fullQueue(reason: String)
}

File: Kotlin/queue.kt
internal class Queue<T>(size: Int) {

  internal var size: Int
    get() = this.size

  internal var isEmpty: Boolean = true
    get() = this.size == 0

  internal var isFull: Boolean = false
    get() = this.size == this.items.size

  private var items: Array<Any?>
  private var first: Int
  private var last: Int

  init {
    this.items = arrayOfNulls<Any?>(size)
    this.last  = size - 1
    this.first = 0
    this.size  = 0
  }

  internal fun peek(): T? {

    if (this.isEmpty) {
      throw Exception("Cannot peek at an empty queue!")
    }

    return this.items[this.first] as T?
  }

  internal fun dequeue(): T? {

    if (this.isEmpty) {
      throw Exception("Cannot dequeue from an empty queue!")
    }

    val item = this.items[this.first]
    this.items[this.first] = null
    this.first = (this.first + 1) % this.items.size
    this.size -= 1
    return item as T?
  }

  internal fun enqueue(item: T?): Unit {

    if (this.isFull) {
      throw Exception("Cannot enqueue into a full queue!")
    }

    this.last = (this.last + 1) % this.items.size
    this.items[this.last] = item
    this.size += 1
  }

}
